src/Polar/kernels.jl

```diff
@@ -404,51 +747,70 @@ KA.@kernel function branch_flow_kernel!(
     slines[ℓ + nlines] = to_flow
 end
 
-function branch_flow_kernel_zygote(
+KA.@kernel function adj_branch_flow_kernel!(
+        adj_slines, vmag, adj_vmag, vang, adj_vang,
         yff_re, yft_re, ytf_re, ytt_re,
         yff_im, yft_im, ytf_im, ytt_im,
-        fr_vmag, to_vmag,
-        cosθ, sinθ
+        f, t, nlines,
    )
+    ℓ = @index(Global, Linear)
+    fr_bus = f[ℓ]
+    to_bus = t[ℓ]
 
-    # branch apparent power limits - from bus
-    yff_abs = yff_re.^2 .+ yff_im.^2
-    yft_abs = yft_re.^2 .+ yft_im.^2
-    yre_fr =   yff_re .* yft_re .+ yff_im .* yft_im
-    yim_fr = .- yff_re .* yft_im .+ yff_im .* yft_re
-
-    fr_flow = fr_vmag.^2 .* (
-        yff_abs .* fr_vmag.^2 .+ yft_abs .* to_vmag.^2 .+
-        2 .* fr_vmag .* to_vmag .* (yre_fr .* cosθ .- yim_fr .* sinθ)
-    )
+    Δθ = vang[fr_bus] - vang[to_bus]
+    cosθ = cos(Δθ)
+    sinθ = sin(Δθ)
 
-    # branch apparent power limits - to bus
-    ytf_abs = ytf_re.^2 + ytf_im.^2
-    ytt_abs = ytt_re.^2 + ytt_im.^2
-    yre_to =   ytf_re .* ytt_re .+ ytf_im .* ytt_im
-    yim_to = - ytf_re .* ytt_im .+ ytf_im .* ytt_re
-
-    to_flow = to_vmag.^2 .* (
-        ytf_abs .* fr_vmag.^2 .+ ytt_abs .* to_vmag.^2 .+
-        2 .* fr_vmag .* to_vmag .* (yre_to .* cosθ .- yim_to .* sinθ)
-    )
-    return vcat(fr_flow, to_flow)
-end
+    # branch apparent power limits - from bus
+    yff_abs = yff_re[ℓ]^2 + yff_im[ℓ]^2
+    yft_abs = yft_re[ℓ]^2 + yft_im[ℓ]^2
+    yre_fr =   yff_re[ℓ] * yft_re[ℓ] + yff_im[ℓ] * yft_im[ℓ]
+    yim_fr = - yff_re[ℓ] * yft_im[ℓ] + yff_im[ℓ] * yft_re[ℓ]
 
-"""
-    accumulate_view(x, vx, indices)
+    # not needed in the reverse run
+    # fr_flow = vmag[fr_bus]^2 * (
+    #     yff_abs * vmag[fr_bus]^2 + yft_abs * vmag[to_bus]^2 +
+    #     2 * vmag[fr_bus] * vmag[to_bus] * (yre_fr * cosθ - yim_fr * sinθ)
+    # )
+    # slines[ℓ] = fr_flow
 
-.+= is broken on views with redundant indices in CUDA.jl leading to a bug Zygote (see #89).
-This implements the .+= operator.
+    # branch apparent power limits - to bus
+    ytf_abs = ytf_re[ℓ]^2 + ytf_im[ℓ]^2
+    ytt_abs = ytt_re[ℓ]^2 + ytt_im[ℓ]^2
+    yre_to =   ytf_re[ℓ] * ytt_re[ℓ] + ytf_im[ℓ] * ytt_im[ℓ]
+    yim_to = - ytf_re[ℓ] * ytt_im[ℓ] + ytf_im[ℓ] * ytt_re[ℓ]
 
-"""
-KA.@kernel function accumulate_view!(x, vx, indices)
-    # This is parallelizable
-    id = @index(Global, Linear)
-    for (j, idx) in enumerate(indices)
-        if id == idx
-            x[id] += vx[j]
-        end
-    end
+    # not needed in the reverse run
+    # to_flow = vmag[to_bus]^2 * (
+    #     ytf_abs * vmag[fr_bus]^2 + ytt_abs * vmag[to_bus]^2 +
+    #     2 * vmag[fr_bus] * vmag[to_bus] * (yre_to * cosθ - yim_to * sinθ)
+    # )
+    # slines[ℓ + nlines] = to_flow
+
+    adj_to_flow = adj_slines[ℓ + nlines]
+    adj_vmag[to_bus] += (2 * vmag[to_bus] * ytf_abs * vmag[fr_bus]^2
```

Comment:
I would replace `2` by `2.0` (same `4` and `6`)

Reply:
You are right. I tried that actually yesterday to make sure that's not the cause of the numerical issues.